v4/export/dump.go

```diff
@@ -196,193 +188,386 @@ func (d *Dumper) Dump() (err error) {
+		zap.String("upper", max.String()))
 
-	if !conf.NoViews {
-		views, err := listAllViews(pool, databases)
-		if err != nil {
-			return err
-		}
-		conf.Tables.Merge(views)
+	count := estimateCount(db, tbl, conn, field, conf)
+	log.Info("get estimated rows count", zap.Uint64("estimateCount", count))
+	if count < conf.Rows {
+		// skip chunk logic if estimates are low
+		log.Debug("skip concurrent dump due to estimate count < rows",
+			zap.Uint64("estimate count", count),
+			zap.Uint64("conf.rows", conf.Rows),
+		)
+		return d.sequentialDumpTable(conn, meta, ir)
 	}
 
-	filterTables(conf)
-	return nil
-}
+	// every chunk would have eventual adjustments
+	estimatedChunks := count / conf.Rows
+	estimatedStep := new(big.Int).Sub(max, min).Uint64()/estimatedChunks + 1
+	bigEstimatedStep := new(big.Int).SetUint64(estimatedStep)
+	cutoff := new(big.Int).Set(min)
 
-func dumpSql(ctx context.Context, conf *Config, connectPool *connectionsPool, writer Writer) error {
-	tableIR, err := SelectFromSql(conf, connectPool.extraConn())
+	selectField, selectLen, err := buildSelectField(conn, db, tbl, conf.CompleteInsert)
 	if err != nil {
 		return err
 	}
 
-	tableDataStartTime := time.Now()
-	err = writer.WriteTableData(ctx, tableIR)
+	orderByClause, err := buildOrderByClause(conf, conn, db, tbl)
 	if err != nil {
-		summary.CollectFailureUnit("dump", err)
 		return err
-	} else {
-		summary.CollectSuccessUnit("dump cost", 1, time.Since(tableDataStartTime))
+	}
+
+	nullValueCondition := fmt.Sprintf("`%s` IS NULL OR ", escapeString(field))
+	for max.Cmp(cutoff) >= 0 {
+		nextCutOff := new(big.Int).Add(cutoff, bigEstimatedStep)
+		where := fmt.Sprintf("%s(`%s` >= %d AND `%s` < %d)", nullValueCondition, escapeString(field), cutoff, escapeString(field), nextCutOff)
+		query := buildSelectQuery(db, tbl, selectField, buildWhereCondition(conf, where), orderByClause)
+		if len(nullValueCondition) > 0 {
+			nullValueCondition = ""
+		}
+		ctxDone := sendDataIRToChan(ctx, newTableData(query, selectLen), ir)
+		if ctxDone {
+			break
+		}
+		cutoff = nextCutOff
 	}
 	return nil
 }
 
-func dumpTable(ctx context.Context, conf *Config, db *sql.Conn, dbName string, table *TableInfo, writer Writer) ([]TableDataIR, error) {
-	tableName := table.Name
-	if !conf.NoSchemas {
-		if table.Type == TableTypeView {
-			viewName := table.Name
-			createTableSQL, createViewSQL, err := ShowCreateView(db, dbName, viewName)
-			if err != nil {
-				return nil, err
-			}
-			return nil, writer.WriteViewMeta(ctx, dbName, viewName, createTableSQL, createViewSQL)
+func sendDataIRToChan(ctx context.Context, ir TableDataIR, irChan chan<- TableDataIR) (ctxDone bool) {
+	for {
+		select {
+		case <-ctx.Done():
+			return true
+		case irChan <- ir:
+			return false
+		default:
+			sleepTime := 1 * time.Second
+			time.Sleep(sleepTime)
+			log.Warn("write channel is full, waiting writer to consume IR",
+				zap.Int("channel size", defaultDumpThreads),
+				zap.Stringer("sleep time", sleepTime))
```

Comment:
I think we don't need this 🤔 If we want to check how many ir are blocked here we can add a metrics and `Inc` before `select`.